recommenders/evaluation/python_evaluation.py

```diff
@@ -743,6 +747,8 @@ def __init__(
                 Interaction here follows the *item choice model* from Castells et al.
             reco_df (pandas.DataFrame): Recommender's prediction output, containing col_user, col_item,
                 col_relevance (optional). Assumed to not contain any duplicate user-item pairs.
+              item_feature_df (pandas.DataFrame): (Optional) It is required only when item_sim_measure='item_feature_vector'. It contains two columns: col_item and features (a feature vector).
+            item_sim_measure (str): (Optional) This column indicates which item similarity measure to be used. Available measures include item_cooccurrence_count (default choice) and item_feature_vector.
             col_user (str): User id column name.
             col_item (str): Item id column name.
             col_relevance (str): This column indicates whether the recommended item is actually
```

Comment:
There is one thing I wanted to discuss with you all @anargyri, @YanZhangADS, @gramhagen, @yueguoguo, @loomlike. 

There is a disparity between the way we are doing rating and ranking metrics, and diversity metrics. In rating and ranking we use functions, whereas in diversity we use a class. I think it makes sense to write a class for pyspark, but I would like to get the views of people on if you guys are ok to do the same in python.

Reply:
Good point. I agree, we should be consistent throughout the file. I think it's better to have them as functions, is there any good reason to have classes?

Reply:
The main advantage of using class is, when one class object is initialized, several commonly shared variables do not need to be calculated multiple times.  e.g. both diversity and serendipity need to use item similarity matrix. By using class, we only need to calculate item similarity matrix once.

Reply:
You can avoid the recomputations without using classes. This is what we do in this file, using LRU cache https://github.com/microsoft/recommenders/blob/9b094e32550adbb326f40b02a7056e179a3e3da0/recommenders/evaluation/python_evaluation.py#L96 
See https://docs.python.org/3/library/functools.html#functools.lru_cache

Reply:
Which suggests that you could use LRU cache functionality to cache the similarity computations.

Reply:
This issue was the topic of some discussion in the past, see e.g. https://github.com/microsoft/recommenders/pull/713

Reply:
This is probably a related issue on optimization, but there is a substantial increase in the runtime of the test for the spark version. 
![image](https://user-images.githubusercontent.com/19534002/133282033-1da51ebf-a13d-402d-bcc2-65b719e75f4e.png)
from
![image](https://user-images.githubusercontent.com/19534002/133282196-6cf2613e-144f-47ff-a080-001737010284.png)

I hope we can reduce this runtime if we a smaller mock movielens dataset as we discussed before instead of the "100k" dataset for the notebook. Would welcome thoughts on other angles of optimization.

Reply:
What do the above screenshots compare? Computation time before and after the similarity based metrics were added?

Reply:
> What do the above screenshots compare? Computation time before and after the similarity based metrics were added?
They were comparing before and after the diversity metrics were added. I will open a different thread to discuss them as they are not related to the changes in this PR. Sorry for confusing the two.